File: src/Wording.Core/Packs/PackIndexReader.cs
using System.Text.Json;
using Wording.Core.Storage;

namespace Wording.Core.Packs;

/// <summary>
/// Parses the published catalogue.
/// <para>
/// A single unusable row is dropped rather than refused. The catalogue is the only way
/// most people will ever find a pack, so one malformed entry must not hide every good one
/// behind an error - unlike a pack itself, where a rejected file simply is not imported
/// and the user loses nothing.
/// </para>
/// </summary>
public static class PackIndexReader
{
    public static IReadOnlyList<PackIndexEntry> Read(ReadOnlySpan<byte> payload)
    {
        if (payload.Length > PackLimits.MaxPayloadBytes)
        {
            throw new WordPackException(
                PackProblem.TooLarge,
                $"the catalogue is {payload.Length} bytes, the limit is {PackLimits.MaxPayloadBytes}");
        }

        PackIndex? index;

        try
        {
            index = JsonSerializer.Deserialize(payload, WordJsonContext.Default.PackIndex);
        }
        catch (JsonException error)
        {
            throw new WordPackException(PackProblem.Malformed, "the catalogue is not valid JSON", error);
        }

        if (index is null)
        {
            throw new WordPackException(PackProblem.Malformed, "the catalogue is empty");
        }

        return Clean(index.Packs);
    }

    /// <summary>Applies every rule, dropping the rows that cannot be shown or fetched.</summary>
    public static IReadOnlyList<PackIndexEntry> Clean(IEnumerable<PackIndexEntry> entries)
    {
        ArgumentNullException.ThrowIfNull(entries);

        var cleaned = new List<PackIndexEntry>();
        var seen = new HashSet<string>(StringComparer.Ordinal);

        foreach (var entry in entries)
        {
            // The identifier decides the address the app will fetch, so it gets exactly
            // the same treatment as one inside a pack.
            if (!PackSlug.TryNormalize(entry.Id, out var slug) || !seen.Add(slug))
            {
                continue;
            }

            var name = Text.Clean(entry.Name);

            if (name.Length == 0)
            {
                continue;
            }

            cleaned.Add(new PackIndexEntry
            {
                Id = slug,
                Name = Text.Truncate(name, PackLimits.MaxNameLength),
                Description = Text.Truncate(Text.Clean(entry.Description ?? string.Empty), PackLimits.MaxDescriptionLength)
                    is { Length: > 0 } text
                    ? text
                    : null,
                Kind = PackKind.Normalize(entry.Kind),
                WordCount = Math.Clamp(entry.WordCount, 0, PackLimits.MaxWords),
            });

            if (cleaned.Count == PackLimits.MaxIndexEntries)
            {
                break;
            }
        }

        return cleaned;
    }
}

File: macos/Sources/WordingKit/PackIndexReader.swift
import Foundation

/// Parses the published catalogue.
///
/// A single unusable row is dropped rather than refused. The catalogue is the only way
/// most people will ever find a pack, so one malformed entry must not hide every good one
/// behind an error - unlike a pack itself, where a rejected file simply is not imported
/// and the user loses nothing.
///
/// A port of `Wording.Core.Packs.PackIndexReader`.
public enum PackIndexReader {
    public static func read(_ payload: Data) throws -> [PackIndexEntry] {
        guard payload.count <= PackLimits.maxPayloadBytes else { throw WordPackError.tooLarge }

        let index: PackIndex

        do {
            index = try JSONDecoder().decode(PackIndex.self, from: payload)
        } catch {
            throw WordPackError.malformed("the catalogue is not valid JSON")
        }

        return clean(index.packs)
    }

    /// Applies every rule, dropping the rows that cannot be shown or fetched.
    public static func clean(_ entries: [PackIndexEntry]) -> [PackIndexEntry] {
        var cleaned: [PackIndexEntry] = []
        var seen = Set<String>()

        for entry in entries {
            // The identifier decides the address the app will fetch, so it gets exactly
            // the same treatment as one inside a pack.
            guard let slug = PackSlug.normalize(entry.id), seen.insert(slug).inserted else { continue }

            let name = PackText.clean(entry.name)

            guard !name.isEmpty else { continue }

            let description = PackText.truncate(
                PackText.clean(entry.description ?? ""),
                to: PackLimits.maxDescriptionLength
            )

            cleaned.append(
                PackIndexEntry(
                    id: slug,
                    name: PackText.truncate(name, to: PackLimits.maxNameLength),
                    description: description.isEmpty ? nil : description,
                    kind: PackKind.normalize(entry.kind),
                    wordCount: min(max(entry.wordCount, 0), PackLimits.maxWords)
                )
            )

            if cleaned.count == PackLimits.maxIndexEntries { break }
        }

        return cleaned
    }
}
